{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c5188904-eec5-475e-b4ab-6a36f27101d3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8391fba0eb6c4cc5a41f45cd5dca1e3e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output()"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "PACKAGES\n",
    "'''\n",
    "#Bio packages for reading protein data\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis as PA\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "import os #File reading and manipulation\n",
    "import pandas as pd #Dataframes to keep track of, and display data\n",
    "import ipywidgets as widgets #Interactable elements\n",
    "from ipywidgets import *\n",
    "import math #Simplifying certain operations\n",
    "\n",
    "pd.set_option('display.max_rows', 150)\n",
    "\n",
    "'''\n",
    "FUNCTIONS\n",
    "'''\n",
    "def fasta_to_dataframe(fastaPath,includeDefaults = True, pH = 7.0, attributes={}): \n",
    "    '''\n",
    "    Reads a FASTA formatted file of a proteome and returns a pandas dataframe recording the attributes of each proteins entry, attributes are adjusted default includes sequence, molecular weight,\n",
    "    ID, and description. All but molecular weight are necessary for rewriting proteins into another FASTA file.\n",
    "    \n",
    "    fastaPath - File path to the FASTA file being read\n",
    "    includeDefautlts - Bool value on whether or not to include the 4 predefined columns\n",
    "    attributes - Columns to go in the dataframe, formatted as a dictionary where each entry key is the column label and the entry is a lambda function that defines the data in that column\n",
    "    '''\n",
    "    ordered_attributes = {}\n",
    "    if includeDefaults:\n",
    "        ordered_attributes['Sequence'] = lambda record : str(record.seq)\n",
    "        ordered_attributes['ID'] = lambda record : record.id\n",
    "        ordered_attributes['Description'] = lambda record : record.description\n",
    "        ordered_attributes['Molecular Weight'] = lambda record : PA(record.seq.replace('X','Q')).molecular_weight()\n",
    "        ordered_attributes['Charge at pH {}'.format(pH)] = lambda record : PA(record.seq.replace('X','Q')).charge_at_pH(pH)\n",
    "    elif not attributes:\n",
    "        raise Exception('No attribute functions, set includeDefault = False or add attribute functions')\n",
    "\n",
    "    for att in attributes.keys():\n",
    "        ordered_attributes[att] = attributes[att]\n",
    "    \n",
    "    #This is built to use the from_dict method to create a dataframe where each dictionary key is a column and it's associated data is in a list bound to that key.\n",
    "    columns = ordered_attributes.keys() #List of our tracked variables for ease of access\n",
    "    dataDict = {} #dictionary the dataframe will be made from\n",
    "    for x in columns: #prep dataDict to have the data added to it, will be a dictionary where each key is tied to an empty list\n",
    "        dataDict[x] = []\n",
    "    with open(fastaPath):\n",
    "        for record in SeqIO.parse(fastaPath,'fasta'):\n",
    "            for attribute in columns: #For each attribute, append the output of it's defined function to the appropriate list in dataDict\n",
    "                dataDict[attribute].append(ordered_attributes[attribute](record)) \n",
    "    df = pd.DataFrame.from_dict(dataDict) #This command builds the dataframe from dataDict\n",
    "    return(df)\n",
    "                \n",
    "def fractionate(df,numFractions,noise=0.0,newIndices = False):\n",
    "    '''\n",
    "    Splits a dataframe into equal segments, returned as a list of sub-dataframes.\n",
    "    \n",
    "    df - Dataframe being split\n",
    "    numFractions - Int value, how many subdivisions are made.\n",
    "    noise - Float value determining how much subdivisions can bleed over into one another. (Ex: noise=0.1 would result in each subdivision growing 10% larger in either \n",
    "            direction, overlapping that data with its neighbors\n",
    "    newIndices - Determines whether the index numbers for the subdivisions will be reset to start from 0 or retain the indices of those entries in the original df\n",
    "    '''\n",
    "    fracLen = round(len(df) / numFractions) #Base length each sub-dataframe will be (excluding the last one)\n",
    "    fuzz = fracLen * noise #How many additional entries will be added as \"noise\"\n",
    "    dfFractions = [] \n",
    "    for n in range(numFractions):\n",
    "        fuzzyMin = max(0,int(round(n*fracLen - fuzz))) #Determine the minimum index for this subdivision\n",
    "        fuzzyMax = min(len(df),int(round((n+1)*fracLen + fuzz))) #Determine the maximum index for this subdivision\n",
    "        if n == numFractions - 1: #check if this fraction is the last one to prevent cutting off data near the end\n",
    "            fuzzyMax = len(df)\n",
    "            \n",
    "        output = df.iloc[fuzzyMin:fuzzyMax]  #This creates a new dataframe for each subdivision so any necessary changes, like adjusting the index numbers, can be made before appending\n",
    "        if newIndices:\n",
    "            output.reset_index(drop=True, inplace=True)\n",
    "            \n",
    "        dfFractions.append(output)\n",
    "    return(dfFractions)\n",
    "\n",
    "def hydroScoreSimple(sequencestr):\n",
    "    '''\n",
    "    Returns a simple score of hydrophobicity by returning the sum of a given sequences residue hydrophobicity values\n",
    "\n",
    "    sequencestr - A Protein Analysis object from biopython\n",
    "    '''\n",
    "    score = 0 #Defines the score at 0\n",
    "    for AA in hydro_dict.keys(): #For each amino acid in the hydrophobicity index, this will check how many are in the sequence and add the product of that and the amino acid's index\n",
    "        AAcount = sequencestr.count_amino_acids()\n",
    "        if AA in AAcount:\n",
    "            score += (sequencestr.count_amino_acids()[AA] * hydro_dict[AA])\n",
    "    return(round(score,2))\n",
    "\n",
    "def hydroScoreSurface(sequencestr,windowSize=2):\n",
    "    '''\n",
    "    Returns a simple score of hydrophobicity by returning the sum of a given sequences residue hydrophobicity values\n",
    "\n",
    "    sequencestr - A Protein Analysis object from biopython\n",
    "    windowSize - Integer value, determines how many amino acids away the program will check. Default is 2.\n",
    "    '''\n",
    "    score = 0 #Defines the score at 0\n",
    "    working_string = sequencestr.sequence\n",
    "    currentAA = 0\n",
    "    while currentAA < len(working_string):\n",
    "        window = working_string[max(0,currentAA - windowSize):min(len(working_string),currentAA+windowSize)] #Grab a window of the surrounding acids from the residue we're observing\n",
    "        windowSurfaceScore = 0\n",
    "        for AA in window: #Totals the surface scores of each residue, check is there because some FASTAs have weird residue entries \n",
    "            if AA in surface_dict.keys():\n",
    "                windowSurfaceScore += surface_dict[AA]\n",
    "        surfaceAvgScore = windowSurfaceScore / len(window) #averages the surface scores of the residues\n",
    "\n",
    "        if working_string[currentAA] in hydro_dict.keys(): #If the aminio acid is valid, add the product of it's hydrophobicity times its surface score\n",
    "            score += hydro_dict[working_string[currentAA]] * surfaceAvgScore\n",
    "\n",
    "        currentAA += 1\n",
    "    return(round(score,2))\n",
    "\n",
    "'''\n",
    "GLOBAL VARIABLES\n",
    "'''\n",
    "hydro_dict = {     #Hydrophobicity values for each amino acid residue, values are relative to Glycine = 0.0, sourced from Monera et al., J. Protein Sci. 1: 319-329 (1995)\n",
    "    #Amino Acids\n",
    "    'A' : 0.41, #Alanine\n",
    "    'C' : 0.49, #Cysteine\n",
    "    'D' : -0.55, #Aspartic acid\n",
    "    'E' : -0.31, #Glutamic acid\n",
    "    'F' : 1.00, #Phenylalanine\n",
    "    'G' : 0.00, #Glycine\n",
    "    'H' : 0.08, #Histidine\n",
    "    'I' : 0.99, #Isoleucine\n",
    "    'K' : -0.23, #Lysine\n",
    "    'L' : 0.97, #Leucine\n",
    "    'M' : 0.74, #Methionine\n",
    "    'N' : -0.28, #Asparagine\n",
    "    'P' : -0.46, #Proline\n",
    "    'Q' : -0.10, #Glutamine\n",
    "    'R' : -0.14, #Arginine\n",
    "    'S' : -0.05, #Serine\n",
    "    'T' : 0.13, #Threonine\n",
    "    'V' : 0.76, #Valine\n",
    "    'W' : 0.97, #Tryptophan\n",
    "    'Y' : 0.63, #Tyrosine\n",
    "    #Variable characters\n",
    "    'B' : -0.415, #Aspartic acid or Asparagine\n",
    "    'J' : 0.98, #Leucine or Isoleucine\n",
    "    'Z' : -0.205 #Glutamic acid or Glutamine\n",
    "}\n",
    "surface_dict = {      \n",
    "    #Likeliness for a residue to be on the surface of a protein. A 2.0 means the residue shows up 2x more than it would by pure random chance\n",
    "    #Sourced from Gallet et al. Journal of Molecular Biology Vol. 302 Issue 4: 917-926 (2000) https://doi.org/10.1006/jmbi.2000.4092\n",
    "    #Amino Acids\n",
    "    'A' : 0.50, #Alanine\n",
    "    'C' : 0.80, #Cysteine\n",
    "    'D' : 1.3, #Aspartic acid\n",
    "    'E' : 1.40, #Glutamic acid\n",
    "    'F' : 0.30, #Phenylalanine\n",
    "    'G' : 0.60, #Glycine\n",
    "    'H' : 1.2, #Histidine\n",
    "    'I' : 0.20, #Isoleucine\n",
    "    'K' : 2.10, #Lysine\n",
    "    'L' : 0.40, #Leucine\n",
    "    'M' : 0.60, #Methionine\n",
    "    'N' : 1.30, #Asparagine\n",
    "    'P' : 0.80, #Proline\n",
    "    'Q' : 1.50, #Glutamine\n",
    "    'R' : 3.80, #Arginine\n",
    "    'S' : 1.00, #Serine\n",
    "    'T' : 0.90, #Threonine\n",
    "    'V' : 0.30, #Valine\n",
    "    'W' : 0.50, #Tryptophan\n",
    "    'Y' : 0.90, #Tyrosine\n",
    "    #Variable characters\n",
    "    'B' : 1.30, #Aspartic acid or Asparagine\n",
    "    'J' : 0.30, #Leucine or Isoleucine\n",
    "    'Z' : 1.45 #Glutamic acid or Glutamine\n",
    "}\n",
    "#These dictionaries are used to link dropdown options to a boolean function that determines whether or not the sequence is separated properly by the media\n",
    "sizeDict = {\n",
    "    #\"Bio-P 0.1-1.8 kDa\" : lambda x : bool(100<x.molecular_weight()<1800),  #Currently disabled because the size range is too small to be useful\n",
    "    \"Bio-P 0.8-4.0 kDa\" : lambda x : bool(800<x.molecular_weight()<4000),\n",
    "    \"Bio-P 1.0-6.0 kDa\" : lambda x : bool(1000<x.molecular_weight()<6000),\n",
    "    \"Bio-P 1.5-20.0 kDa\" : lambda x : bool(1500<x.molecular_weight()<20000),\n",
    "    \"Bio-P 2.5-40.0 kDA\" : lambda x : bool(2500<x.molecular_weight()<40000),\n",
    "    \"Bio-P 3.0-60.0 kDa\" : lambda x : bool(3000<x.molecular_weight()<60000),\n",
    "    \"Bio-P 5.0-100 kDa\" : lambda x : bool(5000<x.molecular_weight()<100000),\n",
    "    \"S-X 0.4-14.0 kDa\" : lambda x : bool(400<x.molecular_weight()<14000),\n",
    "    #\"S-X <2.0 kDA\" : lambda x : bool(0<x.molecular_weight()<2000),         #Currently disabled because the size range is too small to be useful\n",
    "    #\"S-X <0.4 kDA\" : lambda x : bool(0<x.molecular_weight()<400),          #Currently disabled because the size range is too small to be useful\n",
    "    \"Bio-A 10.0 - 500 kDA\" : lambda x : bool(10000<x.molecular_weight()<500000),\n",
    "    \"Bio-A 10.0 - 1500 kDA\" : lambda x : bool(10000<x.molecular_weight()<1500000),\n",
    "}\n",
    "ionDict = {\n",
    "    \"Q Media (Triethylamine +)\" : lambda x : bool(x.charge_at_pH(pHslider.value)<=-0.01),\n",
    "    \"S Media (Sulfite -)\" : lambda x : bool(x.charge_at_pH(pHslider.value)>=0.01),\n",
    "}\n",
    "affinityDict ={\n",
    "    \"Hydrophobic Separation (Simple)\" : lambda x : bool(hydroScoreSimple(x) >= 5.00),   #The cutoff above 5.00 is completely arbitrary for the sake of having something\n",
    "    \"Hydrophobic Separation (Surface Algorithm)\" : lambda x : bool(hydroScoreSurface(x) >= 5.00),   #The cutoff above 5.00 is completely arbitrary for the sake of having something\n",
    "} \n",
    "\n",
    "\n",
    "methodDict = {       #This dictionary holds the previous ones for the method selection menu\n",
    "    'Size Exclusion' : sizeDict,\n",
    "    'Ion Exchange' : ionDict,\n",
    "    'Affinity Chromatography' : affinityDict,\n",
    "}\n",
    "sortColumn = {       #Holds a dictionary matching a filtration method to the variable that proteins should be sorted by in the fractions\n",
    "    'Size Exclusion' : 'Molecular Weight',\n",
    "    'Ion Exchange' : 'Charge at pH 7.0',\n",
    "    'Affinity Chromatography' : 'Affinity Score'\n",
    "}\n",
    "data=os.listdir('data') #List to keep track of previous files for conveinience\n",
    "if '.ipynb_checkpoints' in data:\n",
    "    data.remove('.ipynb_checkpoints')\n",
    "style = {'description_width': 'initial'} #shorthands for widget appearances description width\n",
    "autoLayout = Layout(width='auto')\n",
    "\n",
    "#Output Trackers\n",
    "init2active = False\n",
    "init3active = False\n",
    "\n",
    "inFile = None\n",
    "outFolder = None\n",
    "method = None\n",
    "\n",
    "main_df = pd.DataFrame\n",
    "\n",
    "boxLayout = Layout(display='flex')\n",
    "\n",
    "\n",
    "\n",
    "'''\n",
    "MEDIA & METHOD SELECTION\n",
    "'''\n",
    "def confirmMethod(_): #Swaps media selection to match method\n",
    "    with init:\n",
    "        mediaSelect.options = methodDict[methodSelect.value].keys()\n",
    "        mediaSelect.value = mediaSelect.options[0]\n",
    "\n",
    "def confirmMedia(_):\n",
    "    global method\n",
    "    with init:\n",
    "        method = methodDict[methodSelect.value][mediaSelect.value]\n",
    "def pHUpdate(_):\n",
    "    with init:\n",
    "        sortColumn['Ion Exchange'] = 'Charge at pH {}'.format(round(pHslider.value,2))\n",
    "        \n",
    "init = widgets.Output()\n",
    "init2 = widgets.Output()\n",
    "init3 = widgets.Output()\n",
    "methodSelect = widgets.Dropdown(options=methodDict.keys(),description='Method:')\n",
    "methodSelect.observe(confirmMethod,names='value')\n",
    "mediaSelect = widgets.Dropdown(options=methodDict[methodSelect.value].keys(),description='Media',continuous_updates=False)\n",
    "method = methodDict[methodSelect.value][mediaSelect.value] #Initial declaration of 'method' in case the user doesn't update mediaSelect\n",
    "mediaSelect.observe(confirmMedia,names='value')\n",
    "pHslider = widgets.FloatSlider(value=7.0,min=0,max=14,step=0.1,description='pH',style=style)\n",
    "pHslider.observe(pHUpdate)\n",
    "\n",
    "selectDisplay = Box(children=[Box(children=[methodSelect,pHslider],layout=boxLayout),mediaSelect],layout=Layout(display='flex',flex_flow='column'))\n",
    "\n",
    "'''\n",
    "FASTA INPUT SELECTION\n",
    "'''\n",
    "def confirmInput(_):\n",
    "    global init2active\n",
    "    global inFile\n",
    "    with init:\n",
    "        file = inputFile.value\n",
    "        fileEnd = file.split('.')[1]\n",
    "        fastaEnds = ['fasta', 'fna', 'ffn', 'faa', 'frn', 'fa']\n",
    "        if file == None or file == '':\n",
    "            errorText.value = 'Error: No file selected'\n",
    "        elif fileEnd not in fastaEnds:\n",
    "            errorText.value = 'Error: \\\"{}\\\" is not recognized as a FASTA file'.format(inputFile.value)\n",
    "        else:\n",
    "            if inputFile.value not in data:\n",
    "                previousInputs.append(inputFile.value)\n",
    "            currentInput.value = inputFile.value\n",
    "            inputFile.options = data\n",
    "            errorText.value = ''\n",
    "            inFile = os.path.join('data',inputFile.value)\n",
    "            if not init2active:\n",
    "                display(init2)\n",
    "                init2active = True\n",
    "\n",
    "\n",
    "\n",
    "inputFile = widgets.Combobox(value=None,placeholder='Enter a file to be separated',options=data,description='Unseparated data',style = style,layout = Layout(width='440px'),continuous_update=False)\n",
    "\n",
    "inputButton = widgets.Button(description='Confirm File',layout=Layout(width='auto'))\n",
    "currentInput = widgets.HTML(value='No file selected',description='Current input:',style = style)\n",
    "errorText = widgets.HTML(value=None,description='\\t',style={'text_color':'#CC0000','font_size':'16px'})\n",
    "\n",
    "\n",
    "inputButton.on_click(confirmInput,remove=True)\n",
    "inputFile.observe(confirmInput,names=['value'])\n",
    "\n",
    "\n",
    "\n",
    "inputTop = Box(children=[inputFile,inputButton],layout=boxLayout)\n",
    "inputBottom = Box(children=[currentInput,errorText],layout=boxLayout)\n",
    "inputDisplay=VBox([inputTop,inputBottom])\n",
    "\n",
    "'''\n",
    "OUTPUT FOLDER CREATION\n",
    "'''\n",
    "\n",
    "def checkFolder(_):\n",
    "    global init3active\n",
    "    global outFolder\n",
    "    if os.path.isdir(os.path.join('outputs',outputFile.value)):\n",
    "        with init2:\n",
    "            header.description=f'\\\"{outputFile.value}\\\" already exists. Some files may be replaced. Continue?'\n",
    "            outputFile.disabled = outputButton.disabled = True\n",
    "            display(overwriteCheck)\n",
    "    else:\n",
    "        with init2:\n",
    "            dataFolder = outputFile.value\n",
    "            os.mkdir(os.path.join('outputs',dataFolder))\n",
    "            outputFile.disabled = outputButton.disabled = True\n",
    "            outFolder = os.path.join('outputs',outputFile.value)\n",
    "            writeFASTAs(_)\n",
    "            defPageVars(_)\n",
    "            if not init3active:\n",
    "                display(init3)\n",
    "                init3active = True\n",
    "        \n",
    "\n",
    "def confirmOverwrite(_):\n",
    "    global init3active\n",
    "    global outFolder\n",
    "    with init2:\n",
    "        init2.clear_output()\n",
    "        display(outputDisplay)\n",
    "        outFolder = os.path.join('outputs',outputFile.value)\n",
    "        writeFASTAs(_)\n",
    "        defPageVars(_)\n",
    "        if not init3active:\n",
    "                display(init3)\n",
    "                init3active = True\n",
    "    \n",
    "def denyOverwrite(_):\n",
    "    with init2:\n",
    "        init2.clear_output()\n",
    "        display(outputDisplay)\n",
    "        outputFile.value = ''\n",
    "        outputFile.disabled = outputButton.disabled = False\n",
    "\n",
    "#Generate the output name entry\n",
    "outputFile = widgets.Text(placeholder='Enter a label for the output folder.', style=style,disabled=False,continuous_update=False,)\n",
    "outputButton = widgets.Button(description='Confirm',disabled=False)\n",
    "outputLabel = widgets.Label(value='Select output location:')\n",
    "outputDisplay = VBox([outputLabel,Box(children=[outputFile,outputButton])])\n",
    "\n",
    "#Generate the overwrite prompt\n",
    "confirmButton = widgets.Button(description='Yes',layout=Layout(width='auto'))\n",
    "denyButton = widgets.Button(description='No',layout=Layout(width='auto'))\n",
    "header = widgets.HTML(description=f'\\\"{outputFile.value}\\\" already exists. Continue?',layout=Layout(justify_content='center',width='auto'),style=style)\n",
    "YesNo = Box(children=[confirmButton, denyButton],layout=boxLayout)\n",
    "overwriteCheck = Box([header,YesNo],layout=Layout(display='flex',flex_flow='column',justify_content='center',),)\n",
    "\n",
    "#Define button functionality\n",
    "outputButton.on_click(checkFolder)\n",
    "outputFile.observe(checkFolder,names=['value'])\n",
    "confirmButton.on_click(confirmOverwrite)\n",
    "denyButton.on_click(denyOverwrite)\n",
    "\n",
    "with init2:\n",
    "    display(outputDisplay)\n",
    "\n",
    "    \n",
    "'''\n",
    "WRITING OUTPUT FILES\n",
    "'''\n",
    "\n",
    "#Bind user inputs to named variables for simplicity\n",
    "    #inFile = os.path.join('data',inputFile.value)                                 \n",
    "    #outFolder = os.path.join('outputs',outputFile.value)\n",
    "    #method = methodDict[methodSelect.value][mediaSelect.value]\n",
    "#Create lists of SeqRecord objects that biopython can transform into .fasta files\n",
    "\n",
    "def writeFASTAs(_):\n",
    "    global inFile\n",
    "    global outFolder\n",
    "    global main_df\n",
    "    global methodSelect\n",
    "    if not (inFile and outFolder):\n",
    "        raise Exception('Input/output locations not properly selected')\n",
    "    totals = []\n",
    "    wash = []\n",
    "    separated = []\n",
    "    #Separate proteins by the parameter\n",
    "    with open(inFile):\n",
    "        for record in SeqIO.parse(inFile,'fasta'):\n",
    "            totals.append(record)\n",
    "            sequence = str(record.seq)\n",
    "            #X is a notation for any residue, biopython doesn't accept it though so for the protien analysis we substituted Glutamine since it's close to the average for the properties we care about.\n",
    "            protparams = PA(sequence.replace(\"X\",\"Q\"))\n",
    "            if method(protparams):\n",
    "                separated.append(record)\n",
    "            else:\n",
    "                wash.append(record)\n",
    "\n",
    "    SeqIO.write(totals,os.path.join(outFolder,'{}_total'.format(outputFile.value)),'fasta')\n",
    "    SeqIO.write(separated,os.path.join(outFolder,'{}_separated'.format(outputFile.value)),'fasta')\n",
    "    SeqIO.write(wash,os.path.join(outFolder,'{}_wash'.format(outputFile.value)),'fasta')\n",
    "    \n",
    "    #Define a data frame to easily split into fractions\n",
    "    if methodSelect.value == 'Affinity Chromatography':\n",
    "        additional_attributes = {'Affinity Score' : lambda record : hydroScoreSimple(PA(record.seq.replace('X','Q')))}\n",
    "    else:\n",
    "        additional_attributes = {}\n",
    "\n",
    "    main_df = fasta_to_dataframe(os.path.join(outFolder,'{}_separated'.format(outputFile.value)),includeDefaults=True,pH=round(pHslider.value,2),attributes=additional_attributes)\n",
    "    main_df.index.name = 'Protein Index #'\n",
    "    main_df.sort_values(by=sortColumn[methodSelect.value],inplace=True)\n",
    "    fraction_dfs = fractionate(main_df,7)\n",
    "    for n in range(len(fraction_dfs)):\n",
    "        records = []\n",
    "        for index in range(len(fraction_dfs[n])):\n",
    "            single_frac = fraction_dfs[n]\n",
    "            records.append(SeqRecord(seq = Seq(single_frac.iloc[index]['Sequence']),id=single_frac.iloc[index]['ID'],description=single_frac.iloc[index]['Description']))\n",
    "        SeqIO.write(records,os.path.join(outFolder,'{}_fraction_{}'.format(outputFile.value,n+1)),'fasta')    \n",
    "\n",
    "'''\n",
    "PRESENTING DATA\n",
    "'''\n",
    "def pageUp(_):\n",
    "    with init3:\n",
    "        if (pageNum.value < pageNum.max):\n",
    "            pageNum.value += 1\n",
    "        if pageNum.value == pageNum.max:\n",
    "            nextButton.disabled = True\n",
    "        else:\n",
    "            nextButton.disabled = False\n",
    "        prevButton.disabled = False\n",
    "        \n",
    "        init3.clear_output()\n",
    "        temp_df = main_df.sort_values(by=sortBy.value,ascending=bool(ascButton.icon == 'sort-amount-asc'))\n",
    "        display(dfMenu)\n",
    "        display(temp_df.iloc[(pageNum.value-1)*pageSize.value:pageNum.value*pageSize.value])\n",
    "def pageDown(_):\n",
    "    with init3:\n",
    "        if (pageNum.value > pageNum.min):\n",
    "            pageNum.value -= 1\n",
    "        if pageNum.value == pageNum.min:\n",
    "            prevButton.disabled = True\n",
    "        else:\n",
    "            prevButton.disabled = False\n",
    "        nextButton.disabled = False\n",
    "        \n",
    "        init3.clear_output()\n",
    "        temp_df=main_df.sort_values(by=sortBy.value)\n",
    "        display(dfMenu)\n",
    "        display(temp_df.iloc[(pageNum.value-1)*pageSize.value:pageNum.value*pageSize.value])\n",
    "        \n",
    "def adjustMax(_):\n",
    "    with init3:\n",
    "        pageNum.max = math.ceil(len(main_df)/pageSize.value)\n",
    "        pageNum.value = 1\n",
    "        \n",
    "        init3.clear_output()\n",
    "        temp_df = main_df.sort_values(by=sortBy.value,ascending=bool(ascButton.icon == 'sort-amount-asc'))\n",
    "        display(dfMenu)\n",
    "        display(temp_df.iloc[(pageNum.value-1)*pageSize.value:pageNum.value*pageSize.value])\n",
    "\n",
    "def defPageVars(_):\n",
    "    global sortCategories\n",
    "    sortCategories = [main_df.index.name]\n",
    "    if len(sortCategories) <= 1:\n",
    "        for column in list(main_df.keys()):\n",
    "            sortCategories.append(column)\n",
    "    for i in ['ID','Description','Sequence']: #Define column titles that can be ignored as sortable values\n",
    "        sortCategories.remove(i)\n",
    "    sortBy.options = sortCategories\n",
    "    sortBy.value = sortColumn[methodSelect.value]\n",
    "    \n",
    "    pageNum.max=math.ceil(len(main_df)/pageSize.value)\n",
    "        \n",
    "pageSize = widgets.BoundedIntText(value=25,min=5,max=100,step=5,description='Items per Page',style=style,layout=Layout(width='30%'))\n",
    "pageNum = widgets.BoundedIntText(value=1,min=1,max=1,description='Page:',style=style,layout=Layout(width='auto'))\n",
    "pageSize.observe(adjustMax)\n",
    "\n",
    "\n",
    "#define navigation buttons\n",
    "prevButton = widgets.Button(description='Prev',style=style,layout=Layout(width='auto'),disabled=True)\n",
    "prevButton.on_click(pageDown)\n",
    "nextButton = widgets.Button(description='Next',style=style,layout=Layout(width='auto'))\n",
    "nextButton.on_click(pageUp)\n",
    "\n",
    "#Define a list of categories it makes sense to sort, will automatically add any new columns to list\n",
    "\n",
    "    \n",
    "sortBy = widgets.Dropdown(description='Sort By:')\n",
    "def changeSortBy(_):\n",
    "    with init3:\n",
    "        init3.clear_output()\n",
    "        temp_df = main_df.sort_values(by=sortBy.value,ascending=bool(ascButton.icon == 'sort-amount-asc'))\n",
    "        display(dfMenu)\n",
    "        display(temp_df.iloc[(pageNum.value-1)*pageSize.value:pageNum.value*pageSize.value])\n",
    "\n",
    "sortBy.observe(changeSortBy)\n",
    "ascButton = widgets.Button(icon='sort-amount-asc',layout=Layout(width='auto'))\n",
    "def swapSort(_):\n",
    "    with init3:\n",
    "        if ascButton.icon == 'sort-amount-asc':\n",
    "            ascButton.icon = 'sort-amount-desc'\n",
    "        else:\n",
    "            ascButton.icon = 'sort-amount-asc'\n",
    "        init3.clear_output()\n",
    "        temp_df = main_df.sort_values(by=sortBy.value,ascending=bool(ascButton.icon == 'sort-amount-asc'))\n",
    "        display(dfMenu)\n",
    "        display(temp_df.iloc[(pageNum.value-1)*pageSize.value:pageNum.value*pageSize.value])\n",
    "\n",
    "    \n",
    "ascButton.on_click(swapSort)\n",
    "\n",
    "pageBox = Box(children=[prevButton,pageNum,nextButton,pageSize],layout=boxLayout)\n",
    "dfMenu = Box(children=[pageBox,sortBy,ascButton])\n",
    "\n",
    "with init3:\n",
    "    display(dfMenu)\n",
    "\n",
    "'''\n",
    "INITIALIZATION\n",
    "'''\n",
    "with init:\n",
    "    display(selectDisplay,inputDisplay)\n",
    "init"
   ]
  },
  {
   "cell_type": "raw",
   "id": "7cd3cd4e-20c5-4320-8984-c6bd07a745bd",
   "metadata": {},
   "source": [
    "New naming order\n",
    "1. Full Proteome\n",
    "2. Wash\n",
    "3+. Fractions (MW biggest first, least charged for Ion)\n",
    "Last. Stuff that doesn't separate (MW too small)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
