{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "c5188904-eec5-475e-b4ab-6a36f27101d3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "676923db74e240fb86b5b198eaf01bab",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output()"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "PACKAGES\n",
    "'''\n",
    "#Bio packages for reading protein data\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis as PA\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "import os #File reading and manipulation\n",
    "import pandas as pd #Dataframes to keep track of, and display data\n",
    "import ipywidgets as widgets #Interactable elements\n",
    "from ipywidgets import *\n",
    "import math #Simplifying certain operations\n",
    "\n",
    "pd.set_option('display.max_rows', 150)\n",
    "\n",
    "'''\n",
    "FUNCTIONS\n",
    "'''\n",
    "def fasta_to_dataframe(fastaPath,includeDefaults = True, pH = 7.0, attributes={}): \n",
    "    '''\n",
    "    Reads a FASTA formatted file of a proteome and returns a pandas dataframe recording the attributes of each proteins entry, attributes are adjusted default includes sequence, molecular weight,\n",
    "    ID, and description. All but molecular weight are necessary for rewriting proteins into another FASTA file.\n",
    "    \n",
    "    fastaPath - File path to the FASTA file being read\n",
    "    includeDefautlts - Bool value on whether or not to include the 4 predefined columns\n",
    "    attributes - Columns to go in the dataframe, formatted as a dictionary where each entry key is the column label and the entry is a lambda function that defines the data in that column\n",
    "    '''\n",
    "    attributes.clear()\n",
    "    if includeDefaults:\n",
    "        attributes['Sequence'] = lambda record : str(record.seq)\n",
    "        attributes['ID'] = lambda record : record.id\n",
    "        attributes['Description'] = lambda record : record.description\n",
    "        attributes['Molecular Weight'] = lambda record : PA(record.seq.replace('X','Q')).molecular_weight()\n",
    "        attributes['Charge at pH {}'.format(pH)] = lambda record : PA(record.seq.replace('X','Q')).charge_at_pH(pH)\n",
    "    elif not attributes:\n",
    "        raise Exception('No attribute functions, set includeDefault = False or add attribute functions')\n",
    "    \n",
    "    #This is built to use the from_dict method to create a dataframe where each dictionary key is a column and it's associated data is in a list bound to that key.\n",
    "    columns = attributes.keys() #List of our tracked variables for ease of access\n",
    "    dataDict = {} #dictionary the dataframe will be made from\n",
    "    for x in columns: #prep dataDict to have the data added to it, will be a dictionary where each key is tied to an empty list\n",
    "        dataDict[x] = []\n",
    "    with open(fastaPath):\n",
    "        for record in SeqIO.parse(fastaPath,'fasta'):\n",
    "            for attribute in columns: #For each attribute, append the output of it's defined function to the appropriate list in dataDict\n",
    "                dataDict[attribute].append(attributes[attribute](record)) \n",
    "    df = pd.DataFrame.from_dict(dataDict) #This command builds the dataframe from dataDict\n",
    "    return(df)\n",
    "                \n",
    "def fractionate(df,numFractions,noise=0.0,newIndices = False):\n",
    "    '''\n",
    "    Splits a dataframe into equal segments, returned as a list of sub-dataframes.\n",
    "    \n",
    "    df - Dataframe being split\n",
    "    numFractions - Int value, how many subdivisions are made.\n",
    "    noise - Float value determining how much subdivisions can bleed over into one another. (Ex: noise=0.1 would result in each subdivision growing 10% larger in either \n",
    "            direction, overlapping that data with its neighbors\n",
    "    newIndices - Determines whether the index numbers for the subdivisions will be reset to start from 0 or retain the indices of those entries in the original df\n",
    "    '''\n",
    "    fracLen = round(len(df) / numFractions) #Base length each sub-dataframe will be (excluding the last one)\n",
    "    fuzz = fracLen * noise #How many additional entries will be added as \"noise\"\n",
    "    dfFractions = [] \n",
    "    for n in range(numFractions):\n",
    "        fuzzyMin = max(0,int(round(n*fracLen - fuzz))) #Determine the minimum index for this subdivision\n",
    "        fuzzyMax = min(len(df),int(round((n+1)*fracLen + fuzz))) #Determine the maximum index for this subdivision\n",
    "        if n == numFractions - 1: #check if this fraction is the last one to prevent cutting off data near the end\n",
    "            fuzzyMax = len(df)\n",
    "            \n",
    "        output = df.iloc[fuzzyMin:fuzzyMax]  #This creates a new dataframe for each subdivision so any necessary changes, like adjusting the index numbers, can be made before appending\n",
    "        if newIndices:\n",
    "            output.reset_index(drop=True, inplace=True)\n",
    "            \n",
    "        dfFractions.append(output)\n",
    "    return(dfFractions)\n",
    "\n",
    "\n",
    "\n",
    "'''\n",
    "GLOBAL VARIABLES\n",
    "'''\n",
    "#These dictionaries are used to link dropdown options to a boolean function that determines whether or not the sequence is separated properly by the media\n",
    "sizeDict = {\n",
    "    #\"Bio-P 0.1-1.8 kDa\" : lambda x : bool(100<x.molecular_weight()<1800),  #Currently disabled because the size range is too small to be useful\n",
    "    \"Bio-P 0.8-4.0 kDa\" : lambda x : bool(800<x.molecular_weight()<4000),\n",
    "    \"Bio-P 1.0-6.0 kDa\" : lambda x : bool(1000<x.molecular_weight()<6000),\n",
    "    \"Bio-P 1.5-20.0 kDa\" : lambda x : bool(1500<x.molecular_weight()<20000),\n",
    "    \"Bio-P 2.5-40.0 kDA\" : lambda x : bool(2500<x.molecular_weight()<40000),\n",
    "    \"Bio-P 3.0-60.0 kDa\" : lambda x : bool(3000<x.molecular_weight()<60000),\n",
    "    \"Bio-P 5.0-100 kDa\" : lambda x : bool(5000<x.molecular_weight()<100000),\n",
    "    \"S-X 0.4-14.0 kDa\" : lambda x : bool(400<x.molecular_weight()<14000),\n",
    "    #\"S-X <2.0 kDA\" : lambda x : bool(0<x.molecular_weight()<2000),         #Currently disabled because the size range is too small to be useful\n",
    "    #\"S-X <0.4 kDA\" : lambda x : bool(0<x.molecular_weight()<400),          #Currently disabled because the size range is too small to be useful\n",
    "    \"Bio-A 10.0 - 500 kDA\" : lambda x : bool(10000<x.molecular_weight()<500000),\n",
    "    \"Bio-A 10.0 - 1500 kDA\" : lambda x : bool(10000<x.molecular_weight()<1500000),\n",
    "}\n",
    "ionDict = {\n",
    "    \"Q Media (Triethylamine +)\" : lambda x : bool(x.charge_at_pH(pHslider.value)<=-0.01),\n",
    "    \"S Media (Sulfite -)\" : lambda x : bool(x.charge_at_pH(pHslider.value)>=0.01),\n",
    "}\n",
    "affinityDict ={} #Empty for now until more affinity separations are researched and implemented\n",
    "\n",
    "#This dictionary holds the previous ones for the method selection menu\n",
    "methodDict = {\n",
    "    'Size Exclusion' : sizeDict,\n",
    "    'Ion Exchange' : ionDict,\n",
    "    'Affinity Chromatography' : affinityDict,\n",
    "}\n",
    "sortColumn = {\n",
    "    'Size Exclusion' : 'Molecular Weight',\n",
    "    'Ion Exchange' : 'Charge at pH 7.0',\n",
    "    'Affinity Chromatography' : ''\n",
    "}\n",
    "data=os.listdir('data') #List to keep track of previous files for conveinience\n",
    "data.remove('.ipynb_checkpoints')\n",
    "style = {'description_width': 'initial'} #shorthands for widget appearances description width\n",
    "autoLayout = Layout(width='auto')\n",
    "\n",
    "\n",
    "'''\n",
    "MEDIA & METHOD SELECTION\n",
    "'''\n",
    "def confirmMethod(_): #Swaps media selection to match method\n",
    "    with init:\n",
    "        mediaSelect.options = methodDict[methodSelect.value].keys()\n",
    "        \n",
    "def pHUpdate(_):\n",
    "    with init:\n",
    "        sortColumn['Ion Exchange'] = 'Charge at pH {}'.format(pHslider.value)\n",
    "        \n",
    "init = widgets.Output()\n",
    "methodSelect = widgets.Dropdown(options=methodDict.keys(),description='Method:')\n",
    "methodSelect.observe(confirmMethod)\n",
    "mediaSelect = widgets.Dropdown(options=methodDict[methodSelect.value].keys(),description='Media')\n",
    "pHslider = widgets.FloatSlider(value=7.0,min=0,max=14,step=0.1,description='pH',style=style)\n",
    "pHslider.observe(pHUpdate)\n",
    "selectDisplay = TwoByTwoLayout(top_left=methodSelect,bottom_left=mediaSelect,bottom_right=pHslider,layout=Layout(width='50%'))\n",
    "\n",
    "\n",
    "'''\n",
    "FASTA INPUT SELECTION\n",
    "'''\n",
    "def confirmInput(_):\n",
    "    with init:\n",
    "        file = inputFile.value\n",
    "        fileEnd = file.split('.')[1]\n",
    "        fastaEnds = ['fasta', 'fna', 'ffn', 'faa', 'frn', 'fa']\n",
    "        if fileEnd in fastaEnds:\n",
    "            if inputFile.value not in previousInputs:\n",
    "                previousInputs.append(inputFile.value)\n",
    "            currentInput.value = inputFile.value\n",
    "            inputFile.options = previousInputs\n",
    "            errorText.value = ''\n",
    "        else:\n",
    "            errorText.value = 'Error: \\\"{}\\\" is not recognized as a FASTA file'.format(inputFile.value)\n",
    "\n",
    "boxLayout = Layout(width='30%')\n",
    "\n",
    "inputFile = widgets.Combobox(value=data[0],placeholder='Enter a file to be separated',options=data,description='Unseparated data',style = style)\n",
    "\n",
    "inputButton = widgets.Button(description='Confirm File')\n",
    "currentInput = widgets.HTML(value=inputFile.value,description='Current input:',style = style)\n",
    "errorText = widgets.HTML(value=None,description='\\t',style={'text_color':'#CC0000','font_size':'16px'})\n",
    "\n",
    "\n",
    "inputButton.on_click(confirmInput)\n",
    "inputFile.continuous_update = False\n",
    "inputFile.observe(confirmInput)\n",
    "\n",
    "\n",
    "\n",
    "inputTop = Box(children=[inputFile,inputButton],layout=boxLayout)\n",
    "inputBottom = Box(children=[currentInput,errorText],layout=boxLayout)\n",
    "inputDisplay=VBox([inputTop,inputBottom])\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "'''\n",
    "INITIALIZATION\n",
    "'''\n",
    "with init:\n",
    "    display(selectDisplay,inputDisplay)\n",
    "init"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "194dc54e-afc1-4003-aab9-c873b1be5544",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54c31606-737f-4614-b22a-04dfbdd293db",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d01c617-e763-4766-a05d-62eb8c800739",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
