{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c5188904-eec5-475e-b4ab-6a36f27101d3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e2678b56ec144ac1b1759df8bdfccc2c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output()"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "PACKAGES\n",
    "'''\n",
    "#Bio packages for reading protein data\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis as PA\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "import os #File reading and manipulation\n",
    "import pandas as pd #Dataframes to keep track of, and display data\n",
    "import ipywidgets as widgets #Interactable elements\n",
    "from ipywidgets import *\n",
    "import math #Simplifying certain operations\n",
    "\n",
    "pd.set_option('display.max_rows', 150)\n",
    "\n",
    "'''\n",
    "FUNCTIONS\n",
    "'''\n",
    "def fasta_to_dataframe(fastaPath,includeDefaults = True, pH = 7.0, attributes={}): \n",
    "    '''\n",
    "    Reads a FASTA formatted file of a proteome and returns a pandas dataframe recording the attributes of each proteins entry, attributes are adjusted default includes sequence, molecular weight,\n",
    "    ID, and description. All but molecular weight are necessary for rewriting proteins into another FASTA file.\n",
    "    \n",
    "    fastaPath - File path to the FASTA file being read\n",
    "    includeDefautlts - Bool value on whether or not to include the 4 predefined columns\n",
    "    attributes - Columns to go in the dataframe, formatted as a dictionary where each entry key is the column label and the entry is a lambda function that defines the data in that column\n",
    "    '''\n",
    "    attributes.clear()\n",
    "    if includeDefaults:\n",
    "        attributes['Sequence'] = lambda record : str(record.seq)\n",
    "        attributes['ID'] = lambda record : record.id\n",
    "        attributes['Description'] = lambda record : record.description\n",
    "        attributes['Molecular Weight'] = lambda record : PA(str(record.seq).replace('X','Q')).molecular_weight()\n",
    "        attributes['Charge at pH {}'.format(pH)] = lambda record : PA(str(record.seq).replace('X','Q')).charge_at_pH(pH)\n",
    "    elif not attributes:\n",
    "        raise Exception('No attribute functions, set includeDefault = False or add attribute functions')\n",
    "    \n",
    "    #This is built to use the from_dict method to create a dataframe where each dictionary key is a column and it's associated data is in a list bound to that key.\n",
    "    columns = attributes.keys() #List of our tracked variables for ease of access\n",
    "    dataDict = {} #dictionary the dataframe will be made from\n",
    "    for x in columns: #prep dataDict to have the data added to it, will be a dictionary where each key is tied to an empty list\n",
    "        dataDict[x] = []\n",
    "    with open(fastaPath):\n",
    "        for record in SeqIO.parse(fastaPath,'fasta'):\n",
    "            for attribute in columns: #For each attribute, append the output of it's defined function to the appropriate list in dataDict\n",
    "                dataDict[attribute].append(attributes[attribute](record)) \n",
    "    df = pd.DataFrame.from_dict(dataDict) #This command builds the dataframe from dataDict\n",
    "    return(df)\n",
    "                \n",
    "def fractionate(df,numFractions,noise=0.0,newIndices = False):\n",
    "    '''\n",
    "    Splits a dataframe into equal segments, returned as a list of sub-dataframes.\n",
    "    \n",
    "    df - Dataframe being split\n",
    "    numFractions - Int value, how many subdivisions are made.\n",
    "    noise - Float value determining how much subdivisions can bleed over into one another. (Ex: noise=0.1 would result in each subdivision growing 10% larger in either \n",
    "            direction, overlapping that data with its neighbors\n",
    "    newIndices - Determines whether the index numbers for the subdivisions will be reset to start from 0 or retain the indices of those entries in the original df\n",
    "    '''\n",
    "    fracLen = round(len(df) / numFractions) #Base length each sub-dataframe will be (excluding the last one)\n",
    "    fuzz = fracLen * noise #How many additional entries will be added as \"noise\"\n",
    "    dfFractions = [] \n",
    "    for n in range(numFractions):\n",
    "        fuzzyMin = max(0,int(round(n*fracLen - fuzz))) #Determine the minimum index for this subdivision\n",
    "        fuzzyMax = min(len(df),int(round((n+1)*fracLen + fuzz))) #Determine the maximum index for this subdivision\n",
    "        if n == numFractions - 1: #check if this fraction is the last one to prevent cutting off data near the end\n",
    "            fuzzyMax = len(df)\n",
    "            \n",
    "        output = df.iloc[fuzzyMin:fuzzyMax]  #This creates a new dataframe for each subdivision so any necessary changes, like adjusting the index numbers, can be made before appending\n",
    "        if newIndices:\n",
    "            output.reset_index(drop=True, inplace=True)\n",
    "            \n",
    "        dfFractions.append(output)\n",
    "    return(dfFractions)\n",
    "\n",
    "\n",
    "\n",
    "'''\n",
    "GLOBAL VARIABLES\n",
    "'''\n",
    "#These dictionaries are used to link dropdown options to a boolean function that determines whether or not the sequence is separated properly by the media\n",
    "sizeDict = {\n",
    "    #\"Bio-P 0.1-1.8 kDa\" : lambda x : bool(100<x.molecular_weight()<1800),  #Currently disabled because the size range is too small to be useful\n",
    "    \"Bio-P 0.8-4.0 kDa\" : lambda x : bool(800<x.molecular_weight()<4000),\n",
    "    \"Bio-P 1.0-6.0 kDa\" : lambda x : bool(1000<x.molecular_weight()<6000),\n",
    "    \"Bio-P 1.5-20.0 kDa\" : lambda x : bool(1500<x.molecular_weight()<20000),\n",
    "    \"Bio-P 2.5-40.0 kDA\" : lambda x : bool(2500<x.molecular_weight()<40000),\n",
    "    \"Bio-P 3.0-60.0 kDa\" : lambda x : bool(3000<x.molecular_weight()<60000),\n",
    "    \"Bio-P 5.0-100 kDa\" : lambda x : bool(5000<x.molecular_weight()<100000),\n",
    "    \"S-X 0.4-14.0 kDa\" : lambda x : bool(400<x.molecular_weight()<14000),\n",
    "    #\"S-X <2.0 kDA\" : lambda x : bool(0<x.molecular_weight()<2000),         #Currently disabled because the size range is too small to be useful\n",
    "    #\"S-X <0.4 kDA\" : lambda x : bool(0<x.molecular_weight()<400),          #Currently disabled because the size range is too small to be useful\n",
    "    \"Bio-A 10.0 - 500 kDA\" : lambda x : bool(10000<x.molecular_weight()<500000),\n",
    "    \"Bio-A 10.0 - 1500 kDA\" : lambda x : bool(10000<x.molecular_weight()<1500000),\n",
    "}\n",
    "ionDict = {\n",
    "    \"Q Media (Triethylamine +)\" : lambda x : bool(x.charge_at_pH(pHslider.value)<=-0.01),\n",
    "    \"S Media (Sulfite -)\" : lambda x : bool(x.charge_at_pH(pHslider.value)>=0.01),\n",
    "}\n",
    "affinityDict ={} #Empty for now until more affinity separations are researched and implemented\n",
    "\n",
    "\n",
    "methodDict = {       #This dictionary holds the previous ones for the method selection menu\n",
    "    'Size Exclusion' : sizeDict,\n",
    "    'Ion Exchange' : ionDict,\n",
    "    #'Affinity Chromatography' : affinityDict,                      #Commented out until there's a algoritm implemented\n",
    "}\n",
    "sortColumn = {       #Holds a dictionary matching a filtration method to the variable that proteins should be sorted by in the fractions\n",
    "    'Size Exclusion' : 'Molecular Weight',\n",
    "    'Ion Exchange' : 'Charge at pH 7.0',\n",
    "    'Affinity Chromatography' : ''\n",
    "}\n",
    "data=os.listdir('data') #List to keep track of previous files for conveinience\n",
    "if '.ipynb_checkpoints' in data:\n",
    "    data.remove('.ipynb_checkpoints')\n",
    "style = {'description_width': 'initial'} #shorthands for widget appearances description width\n",
    "autoLayout = Layout(width='auto')\n",
    "\n",
    "#Output Trackers\n",
    "init2active = False\n",
    "init3active = False\n",
    "\n",
    "inFile = None\n",
    "outFolder = None\n",
    "method = None\n",
    "\n",
    "main_df = pd.DataFrame\n",
    "\n",
    "boxLayout = Layout(display='flex')\n",
    "\n",
    "\n",
    "\n",
    "'''\n",
    "MEDIA & METHOD SELECTION\n",
    "'''\n",
    "def confirmMethod(_): #Swaps media selection to match method\n",
    "    with init:\n",
    "        mediaSelect.options = methodDict[methodSelect.value].keys()\n",
    "        mediaSelect.value = mediaSelect.options[0]\n",
    "\n",
    "def confirmMedia(_):\n",
    "    global method\n",
    "    with init:\n",
    "        method = methodDict[methodSelect.value][mediaSelect.value]\n",
    "        \n",
    "def pHUpdate(_):\n",
    "    with init:\n",
    "        sortColumn['Ion Exchange'] = 'Charge at pH {}'.format(pHslider.value)\n",
    "        \n",
    "init = widgets.Output()\n",
    "init2 = widgets.Output()\n",
    "init3 = widgets.Output()\n",
    "dfOutput = widgets.Output()\n",
    "\n",
    "methodSelect = widgets.Dropdown(options=methodDict.keys(),description='Method:')\n",
    "methodSelect.observe(confirmMethod,names='value')\n",
    "mediaSelect = widgets.Dropdown(options=methodDict[methodSelect.value].keys(),description='Media',continuous_updates=False)\n",
    "method = methodDict[methodSelect.value][mediaSelect.value] #Initial declaration of 'method' in case the user doesn't update mediaSelect\n",
    "mediaSelect.observe(confirmMedia,names='value')\n",
    "pHslider = widgets.FloatSlider(value=7.0,min=0,max=14,step=0.1,description='pH',style=style)\n",
    "pHslider.observe(pHUpdate)\n",
    "#selectDisplay = TwoByTwoLayout(top_left=methodSelect,bottom_left=mediaSelect,bottom_right=pHslider,layout=Layout())\n",
    "selectDisplay = Box(children=[Box(children=[methodSelect,mediaSelect],layout=Layout(display='flex',flex_flow='column')),pHslider],layout=boxLayout)\n",
    "\n",
    "'''\n",
    "FASTA INPUT SELECTION\n",
    "'''\n",
    "def confirmInput(_):\n",
    "    global init2active\n",
    "    global inFile\n",
    "    with init:\n",
    "        file = inputFile.value\n",
    "        fileEnd = file.split('.')[1]\n",
    "        fastaEnds = ['fasta', 'fna', 'ffn', 'faa', 'frn', 'fa']\n",
    "        if file == None or file == '':\n",
    "            errorText.value = 'Error: No file selected'\n",
    "        elif fileEnd not in fastaEnds:\n",
    "            errorText.value = 'Error: \\\"{}\\\" is not recognized as a FASTA file'.format(inputFile.value)\n",
    "        else:\n",
    "            if inputFile.value not in data:\n",
    "                previousInputs.append(inputFile.value)\n",
    "            currentInput.value = inputFile.value\n",
    "            inputFile.options = data\n",
    "            errorText.value = ''\n",
    "            inFile = os.path.join('data',inputFile.value)\n",
    "            if not init2active:\n",
    "                display(init2)\n",
    "                init2active = True\n",
    "\n",
    "\n",
    "\n",
    "inputFile = widgets.Combobox(value=None,placeholder='Enter a file to be separated',options=data,description='Unseparated data',style = style,layout = Layout(width='440px'),continuous_update=False)\n",
    "\n",
    "inputButton = widgets.Button(description='Confirm File')\n",
    "currentInput = widgets.HTML(value='No file selected',description='Current input:',style = style)\n",
    "errorText = widgets.HTML(value=None,description='\\t',style={'text_color':'#CC0000','font_size':'16px'})\n",
    "\n",
    "\n",
    "inputButton.on_click(confirmInput,remove=True)\n",
    "inputFile.observe(confirmInput,names=['value'])\n",
    "\n",
    "\n",
    "\n",
    "inputTop = Box(children=[inputFile,inputButton],layout=boxLayout)\n",
    "inputBottom = Box(children=[currentInput,errorText],layout=boxLayout)\n",
    "inputDisplay=Box(children=[inputTop,inputBottom],layout=Layout(display='flex',flex_flow='column'))\n",
    "\n",
    "'''\n",
    "OUTPUT FOLDER CREATION\n",
    "'''\n",
    "\n",
    "def checkFolder(_):\n",
    "    global init3active\n",
    "    global outFolder\n",
    "    if os.path.isdir(os.path.join('outputs',outputFile.value)):\n",
    "        with init2:\n",
    "            header.description=f'\\\"{outputFile.value}\\\" already exists. Some files may be replaced. Continue?'\n",
    "            outputFile.disabled = outputButton.disabled = True\n",
    "            display(overwriteCheck)\n",
    "    else:\n",
    "        with init2:\n",
    "            dataFolder = outputFile.value\n",
    "            os.mkdir(os.path.join('outputs',dataFolder))\n",
    "            outputFile.disabled = outputButton.disabled = True\n",
    "            outFolder = os.path.join('outputs',outputFile.value)\n",
    "            writeFASTAs(_)\n",
    "            if not init3active:\n",
    "                with dfOutput:\n",
    "                    dfOutput.clear_output()\n",
    "                    display(main_df)\n",
    "                defPageVars\n",
    "                display(init3)\n",
    "                init3active = True\n",
    "        \n",
    "\n",
    "def confirmOverwrite(_):\n",
    "    global init3active\n",
    "    global outFolder\n",
    "    with init2:\n",
    "        init2.clear_output()\n",
    "        display(outputDisplay)\n",
    "        outFolder = os.path.join('outputs',outputFile.value)\n",
    "        writeFASTAs(_)\n",
    "        if not init3active:\n",
    "            defPageVars\n",
    "            display(init3)\n",
    "            with dfOutput:\n",
    "                dfOutput.clear_output()\n",
    "                display(main_df)\n",
    "            init3active = True\n",
    "    \n",
    "def denyOverwrite(_):\n",
    "    with init2:\n",
    "        init2.clear_output()\n",
    "        display(outputDisplay)\n",
    "        outputFile.value = ''\n",
    "        outputFile.disabled = outputButton.disabled = False\n",
    "\n",
    "#Generate the output name entry\n",
    "outputFile = widgets.Text(placeholder='Enter a label for the output folder.', style=style,disabled=False,continuous_update=False,)\n",
    "outputButton = widgets.Button(description='Confirm',disabled=False)\n",
    "outputLabel = widgets.Label(value='Select output location:')\n",
    "outputDisplay = Box(children=[outputLabel,Box(children=[outputFile,outputButton])],layout=Layout(display='flex',flex_flow='column'))\n",
    "\n",
    "#Generate the overwrite prompt\n",
    "confirmButton = widgets.Button(description='Yes',layout=Layout(width='auto'))\n",
    "denyButton = widgets.Button(description='No',layout=Layout(width='auto'))\n",
    "header = widgets.HTML(description=f'\\\"{outputFile.value}\\\" already exists. Continue?',layout=Layout(justify_content='center',width='auto'),style=style)\n",
    "YesNo = Box(children=[confirmButton, denyButton],layout=Layout())\n",
    "overwriteCheck = Box(children=[header,YesNo],layout=Layout(justify_content='center',flex_flow='column'),)\n",
    "\n",
    "#Define button functionality\n",
    "outputButton.on_click(checkFolder)\n",
    "outputFile.observe(checkFolder,names=['value'])\n",
    "confirmButton.on_click(confirmOverwrite)\n",
    "denyButton.on_click(denyOverwrite)\n",
    "\n",
    "with init2:\n",
    "    display(outputDisplay)\n",
    "\n",
    "    \n",
    "'''\n",
    "WRITING OUTPUT FILES\n",
    "'''\n",
    "\n",
    "#Bind user inputs to named variables for simplicity\n",
    "    #inFile = os.path.join('data',inputFile.value)                                 \n",
    "    #outFolder = os.path.join('outputs',outputFile.value)\n",
    "    #method = methodDict[methodSelect.value][mediaSelect.value]\n",
    "#Create lists of SeqRecord objects that biopython can transform into .fasta files\n",
    "\n",
    "def writeFASTAs(_):\n",
    "    global inFile\n",
    "    global outFolder\n",
    "    global main_df\n",
    "    if not (inFile and outFolder):\n",
    "        raise Exception('Input/output locations not properly selected')\n",
    "    totals = []\n",
    "    wash = []\n",
    "    separated = []\n",
    "    #Separate proteins by the parameter\n",
    "    with open(inFile):\n",
    "        for record in SeqIO.parse(inFile,'fasta'):\n",
    "            totals.append(record)\n",
    "            sequence = str(record.seq)\n",
    "            #X is a notation for any residue, biopython doesn't accept it though so for the protien analysis we substituted Glutamine since it's close to the average for the properties we care about.\n",
    "            protparams = PA(sequence.replace(\"X\",\"Q\"))\n",
    "            if method(protparams):\n",
    "                separated.append(record)\n",
    "            else:\n",
    "                wash.append(record)\n",
    "\n",
    "    SeqIO.write(totals,os.path.join(outFolder,'{}_total'.format(outputFile.value)),'fasta')\n",
    "    SeqIO.write(separated,os.path.join(outFolder,'{}_separated'.format(outputFile.value)),'fasta')\n",
    "    SeqIO.write(wash,os.path.join(outFolder,'{}_wash'.format(outputFile.value)),'fasta')\n",
    "    \n",
    "    #Define a data frame to easily split into fractions\n",
    "    main_df = fasta_to_dataframe(os.path.join(outFolder,'{}_separated'.format(outputFile.value)),includeDefaults=True,pH=pHslider.value)\n",
    "    main_df.index.name = 'Protein Index #'\n",
    "    main_df.sort_values(by=sortColumn[methodSelect.value],inplace=True)\n",
    "    fraction_dfs = fractionate(main_df,7)\n",
    "    for n in range(len(fraction_dfs)):\n",
    "        records = []\n",
    "        for index in range(len(fraction_dfs[n])):\n",
    "            single_frac = fraction_dfs[n]\n",
    "            records.append(SeqRecord(seq = Seq(single_frac.iloc[index]['Sequence']),id=single_frac.iloc[index]['ID'],description=single_frac.iloc[index]['Description']))\n",
    "        SeqIO.write(records,os.path.join(outFolder,'{}_fraction_{}'.format(outputFile.value,n+1)),'fasta')    \n",
    "\n",
    "'''\n",
    "PRESENTING DATA\n",
    "'''\n",
    "def pageUp(_):\n",
    "    with init3:\n",
    "        if (pageNum.value < pageNum.max):\n",
    "            pageNum.value += 1\n",
    "        if pageNum.value == pageNum.max:\n",
    "            nextButton.disabled = True\n",
    "        else:\n",
    "            nextButton.disabled = False\n",
    "        prevButton.disabled = False\n",
    "        \n",
    "    with dfOutput:\n",
    "        temp_df = main_df.sort_values(by=sortBy.value,ascending=bool(ascButton.icon == 'sort-amount-asc'))\n",
    "        display(temp_df.iloc[(pageNum.value-1)*pageSize.value:pageNum.value*pageSize.value])\n",
    "def pageDown(_):\n",
    "    with init3:\n",
    "        if (pageNum.value > pageNum.min):\n",
    "            pageNum.value -= 1\n",
    "        if pageNum.value == pageNum.min:\n",
    "            prevButton.disabled = True\n",
    "        else:\n",
    "            prevButton.disabled = False\n",
    "        nextButton.disabled = False\n",
    "        \n",
    "    with dfOutput:\n",
    "        temp_df = main_df.sort_values(by=sortBy.value,ascending=bool(ascButton.icon == 'sort-amount-asc'))\n",
    "        display(temp_df.iloc[(pageNum.value-1)*pageSize.value:pageNum.value*pageSize.value])\n",
    "        \n",
    "def adjustMax(_):\n",
    "    with init3:\n",
    "        pageNum.max = math.ceil(len(main_df)/pageSize.value)\n",
    "        pageNum.value = 1\n",
    "        \n",
    "    with dfOutput:\n",
    "        temp_df = main_df.sort_values(by=sortBy.value,ascending=bool(ascButton.icon == 'sort-amount-asc'))\n",
    "        display(temp_df.iloc[(pageNum.value-1)*pageSize.value:pageNum.value*pageSize.value])\n",
    "        \n",
    "def defPageVars(_):\n",
    "    global sortCategories\n",
    "    sortCategories = [main_df.index.name]\n",
    "    for column in list(main_df.keys()):\n",
    "        if column not in sortCategories:\n",
    "            sortCategories.append(column)\n",
    "    for i in ['ID','Description','Sequence']: #Define column titles that can be ignored as sortable values\n",
    "        sortCategories.remove(i)\n",
    "    sortBy.options = sortCategories\n",
    "    sortBy.value = sortColumn[methodSelect.value]\n",
    "    pageNum.max=math.ceil(len(main_df)/pageSize.value)\n",
    "        \n",
    "pageSize = widgets.BoundedIntText(value=25,min=5,max=100,step=5,description='Items per Page',style=style,layout=Layout(width='30%'))\n",
    "pageNum = widgets.BoundedIntText(value=1,min=1,max=1,description='Page:',style=style,layout=Layout(width='18%'))\n",
    "pageSize.observe(adjustMax)\n",
    "\n",
    "\n",
    "#define navigation buttons\n",
    "prevButton = widgets.Button(description='Prev',style=style,layout=Layout(width='10%'),disabled=True)\n",
    "prevButton.on_click(pageDown)\n",
    "nextButton = widgets.Button(description='Next',style=style,layout=Layout(width='10%'))\n",
    "nextButton.on_click(pageUp)\n",
    "\n",
    "#Define a list of categories it makes sense to sort, will automatically add any new columns to list\n",
    "\n",
    "    \n",
    "sortBy = widgets.Dropdown(description='Sort By:')\n",
    "def changeSortBy(_):\n",
    "    with init3:\n",
    "        init3.clear_output()\n",
    "        temp_df = main_df.sort_values(by=sortBy.value,ascending=bool(ascButton.icon == 'sort-amount-asc'))\n",
    "        display(dfMenu)\n",
    "        display(temp_df.iloc[(pageNum.value-1)*pageSize.value:pageNum.value*pageSize.value])\n",
    "\n",
    "sortBy.observe(changeSortBy)\n",
    "ascButton = widgets.Button(icon='sort-amount-asc',layout=Layout(max_width='35px'))\n",
    "def swapSort(_):\n",
    "    with init3:\n",
    "        if ascButton.icon == 'sort-amount-asc':\n",
    "            ascButton.icon = 'sort-amount-desc'\n",
    "        else:\n",
    "            ascButton.icon = 'sort-amount-asc'\n",
    "    with dfOutput:\n",
    "        temp_df = main_df.sort_values(by=sortBy.value,ascending=bool(ascButton.icon == 'sort-amount-asc'))\n",
    "        display(temp_df.iloc[(pageNum.value-1)*pageSize.value:pageNum.value*pageSize.value])\n",
    "    \n",
    "ascButton.on_click(swapSort)\n",
    "\n",
    "pageBox = Box(children=[prevButton,pageNum,nextButton,pageSize],layout=boxLayout)\n",
    "dfMenu = Box(children=[pageBox,sortBy,ascButton])\n",
    "\n",
    "with init3:\n",
    "    display(dfMenu)\n",
    "    display(dfOutput)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "'''\n",
    "INITIALIZATION\n",
    "'''\n",
    "with init:\n",
    "    display(selectDisplay,inputDisplay)\n",
    "init"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8d01c617-e763-4766-a05d-62eb8c800739",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pd.set_option('display.max_rows', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "36c968e6-45e0-4005-a7c2-81ba00638509",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with dfOutput:\n",
    "    display(main_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "63ad1d0c-e076-4f9f-a1a6-1a242c3fc4cb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "01b32bd0ccae4b65b3388eb7907bf340",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Dropdown(description='Sort By:', index=1, options=('Protein Index #', 'Molecular Weight', 'Charge at pH 7.0'),…"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sortBy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b0466e2b-e8b2-4be4-b362-bb9839cc058b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "defPageVars(_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1f9f55dc-9f04-49fd-8775-13594b14a2e4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2282fff0a57f469796bbfbf35e1e8dad",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output(outputs=({'output_type': 'display_data', 'data': {'text/plain': \"                                      …"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfOutput"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "199aa4bc-2e19-4cfd-be49-b474c76eacd8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "idx2 = main_df.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4657c696-699f-48b0-ba38-a3445b2dfe9f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['A', 'C', 'A', 'B'], dtype='object', name='score')"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idx = pd.Index(['A', 'C', 'A', 'B'], name='score')\n",
    "idx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1c719427-714a-454b-9600-c8963a00a82b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Int64Index([232, 135,  57, 116, 148, 265, 175, 102,  55, 124,\n",
       "            ...\n",
       "             96, 138, 107,  21,  40,   1,  25, 130,  50,  46],\n",
       "           dtype='int64', name='Protein Index #', length=282)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idx2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "8dc139de-ea06-4fc8-ade5-e6259223023d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "25be011e-c236-4eb9-8de0-b770f5bc2535",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>ID</th>\n",
       "      <th>Description</th>\n",
       "      <th>Molecular Weight</th>\n",
       "      <th>Charge at pH 7.0</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Protein Index #</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>MSRLFFVCRKVR</td>\n",
       "      <td>sp|P17776|YZPY_ECOLX</td>\n",
       "      <td>sp|P17776|YZPY_ECOLX Putative PyrE leader pept...</td>\n",
       "      <td>1541.9281</td>\n",
       "      <td>3.489425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>XNNIVFYSLGNVNSYQGG</td>\n",
       "      <td>sp|P20860|FMF1_ECOLX</td>\n",
       "      <td>sp|P20860|FMF1_ECOLX F7-1 fimbrial protein (Fr...</td>\n",
       "      <td>1974.0900</td>\n",
       "      <td>-0.241896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>NTFYCCELCCNPACAGCY</td>\n",
       "      <td>sp|P01560|HSTB_ECOLX</td>\n",
       "      <td>sp|P01560|HSTB_ECOLX Heat-stable enterotoxin S...</td>\n",
       "      <td>1978.2959</td>\n",
       "      <td>-1.298492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>MDAVLSVQGASAPVKKKSFFS</td>\n",
       "      <td>sp|P13948|PIL3_ECOLX</td>\n",
       "      <td>sp|P13948|PIL3_ECOLX Pilin (Fragment) OS=Esche...</td>\n",
       "      <td>2197.5524</td>\n",
       "      <td>1.498478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>MKEPKSNIRKLIDIGGLIGR</td>\n",
       "      <td>sp|P33789|TRAY8_ECOLX</td>\n",
       "      <td>sp|P33789|TRAY8_ECOLX Relaxosome protein TraY ...</td>\n",
       "      <td>2238.6956</td>\n",
       "      <td>2.501269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MNINKPLEILGHVSWLWASSPLHRNWPVSLFAINVLPAIQANQYVL...</td>\n",
       "      <td>sp|P06736|HLYC_ECOLX</td>\n",
       "      <td>sp|P06736|HLYC_ECOLX Protein-lysine myristoylt...</td>\n",
       "      <td>19769.3792</td>\n",
       "      <td>2.027262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>MDTTQVTLIHKILAAADERNLPLWIGGGWAIDARLGRVTRKHDDID...</td>\n",
       "      <td>sp|P0AE04|AADB_ECOLX</td>\n",
       "      <td>sp|P0AE04|AADB_ECOLX 2''-aminoglycoside nucleo...</td>\n",
       "      <td>19873.1785</td>\n",
       "      <td>-16.218067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>RFAHDPMAGGHRMWQMAGLKAQRAQTDVNNKQAAFDAAAKEKADAD...</td>\n",
       "      <td>sp|P18000|CEA5_ECOLX</td>\n",
       "      <td>sp|P18000|CEA5_ECOLX Colicin-E5 (Fragment) OS=...</td>\n",
       "      <td>19873.8051</td>\n",
       "      <td>3.955263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>MASLNMIVAVNKTGGIGFENQIPWHEPEDLKHFKAVTMNSVLIMGR...</td>\n",
       "      <td>sp|Q59397|DYR9_ECOLX</td>\n",
       "      <td>sp|Q59397|DYR9_ECOLX Dihydrofolate reductase t...</td>\n",
       "      <td>19878.7806</td>\n",
       "      <td>0.767784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>MKKLAIVFTMLLIAGCSSSQDSANNQIDELGKENNSLFTFRNIQSG...</td>\n",
       "      <td>sp|Q46670|CDTC_ECOLX</td>\n",
       "      <td>sp|Q46670|CDTC_ECOLX Cytolethal distending tox...</td>\n",
       "      <td>19880.7241</td>\n",
       "      <td>-0.179342</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>282 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                          Sequence  \\\n",
       "Protein Index #                                                      \n",
       "232                                                   MSRLFFVCRKVR   \n",
       "135                                             XNNIVFYSLGNVNSYQGG   \n",
       "57                                              NTFYCCELCCNPACAGCY   \n",
       "116                                          MDAVLSVQGASAPVKKKSFFS   \n",
       "148                                           MKEPKSNIRKLIDIGGLIGR   \n",
       "...                                                            ...   \n",
       "1                MNINKPLEILGHVSWLWASSPLHRNWPVSLFAINVLPAIQANQYVL...   \n",
       "25               MDTTQVTLIHKILAAADERNLPLWIGGGWAIDARLGRVTRKHDDID...   \n",
       "130              RFAHDPMAGGHRMWQMAGLKAQRAQTDVNNKQAAFDAAAKEKADAD...   \n",
       "50               MASLNMIVAVNKTGGIGFENQIPWHEPEDLKHFKAVTMNSVLIMGR...   \n",
       "46               MKKLAIVFTMLLIAGCSSSQDSANNQIDELGKENNSLFTFRNIQSG...   \n",
       "\n",
       "                                    ID  \\\n",
       "Protein Index #                          \n",
       "232               sp|P17776|YZPY_ECOLX   \n",
       "135               sp|P20860|FMF1_ECOLX   \n",
       "57                sp|P01560|HSTB_ECOLX   \n",
       "116               sp|P13948|PIL3_ECOLX   \n",
       "148              sp|P33789|TRAY8_ECOLX   \n",
       "...                                ...   \n",
       "1                 sp|P06736|HLYC_ECOLX   \n",
       "25                sp|P0AE04|AADB_ECOLX   \n",
       "130               sp|P18000|CEA5_ECOLX   \n",
       "50                sp|Q59397|DYR9_ECOLX   \n",
       "46                sp|Q46670|CDTC_ECOLX   \n",
       "\n",
       "                                                       Description  \\\n",
       "Protein Index #                                                      \n",
       "232              sp|P17776|YZPY_ECOLX Putative PyrE leader pept...   \n",
       "135              sp|P20860|FMF1_ECOLX F7-1 fimbrial protein (Fr...   \n",
       "57               sp|P01560|HSTB_ECOLX Heat-stable enterotoxin S...   \n",
       "116              sp|P13948|PIL3_ECOLX Pilin (Fragment) OS=Esche...   \n",
       "148              sp|P33789|TRAY8_ECOLX Relaxosome protein TraY ...   \n",
       "...                                                            ...   \n",
       "1                sp|P06736|HLYC_ECOLX Protein-lysine myristoylt...   \n",
       "25               sp|P0AE04|AADB_ECOLX 2''-aminoglycoside nucleo...   \n",
       "130              sp|P18000|CEA5_ECOLX Colicin-E5 (Fragment) OS=...   \n",
       "50               sp|Q59397|DYR9_ECOLX Dihydrofolate reductase t...   \n",
       "46               sp|Q46670|CDTC_ECOLX Cytolethal distending tox...   \n",
       "\n",
       "                 Molecular Weight  Charge at pH 7.0  \n",
       "Protein Index #                                      \n",
       "232                     1541.9281          3.489425  \n",
       "135                     1974.0900         -0.241896  \n",
       "57                      1978.2959         -1.298492  \n",
       "116                     2197.5524          1.498478  \n",
       "148                     2238.6956          2.501269  \n",
       "...                           ...               ...  \n",
       "1                      19769.3792          2.027262  \n",
       "25                     19873.1785        -16.218067  \n",
       "130                    19873.8051          3.955263  \n",
       "50                     19878.7806          0.767784  \n",
       "46                     19880.7241         -0.179342  \n",
       "\n",
       "[282 rows x 5 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "main_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "678efe23-e296-4fdf-b142-484b989f908c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
